keras/layers/pooling/base_generalized_pooling2d.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Private base class for generalized pooling 2D layers."""
+
+
+from keras import backend
+from keras.engine.base_layer import Layer
+from keras.engine.input_spec import InputSpec
+from keras.utils import conv_utils
+
+
+class GeneralizedPooling2D(Layer):
```

Comment:
Do we really need separate 1D/2D/3D versions? why not a single layer that adapts to its input dimensionality?

Reply:
I wanted to make it consistent with other [pooling layer](https://github.com/keras-team/keras/blob/master/keras/layers/pooling/__init__.py#L20-L43), the way they are written. 

Or, I should use a single base layer with `build` method that will adapt to its input shape?

```python
class base_gem_pool(layers.Layer):
	def build(self, input_shape):
		if len(input_shape) == 3:
 			# config for 1D
		if len(input_shape) == 4:
			# config for 2D 
	....
```

Reply:
> Or, I should use a single base layer with build method that will adapt to its input shape?

That sounds good.

IMO having a single layer is doable and is simpler. The logic is easier to factor in this way than for the pooling layers.